File: Dec10Project/app/src/main/java/com/example/dec10project/MainActivity.java
package com.example.dec10project;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RelativeLayout;
import android.widget.Switch;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

public class MainActivity extends AppCompatActivity {
    EditText nameText;
    EditText yearText;
    TextView msgText;
    Button addButton;
    Switch stdSwitch;
    RadioButton redradioBut;
    RadioButton greenradioBut;
    RadioButton blueradioBut;


    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Log.d("my app","in onCreate Function");
        nameText = findViewById(R.id.nametext);
        yearText = findViewById(R.id.yeartext);
        msgText = findViewById(R.id.welcomeText);
        addButton = findViewById(R.id.addBut);
        stdSwitch = findViewById(R.id.stdSwitch);
        redradioBut = findViewById(R.id.red);
        greenradioBut = findViewById(R.id.green);
        blueradioBut = findViewById(R.id.blue);


        redradioBut.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getWindow().getDecorView().setBackgroundColor(getColor(R.color.red));
            }
        });

        greenradioBut.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getWindow().getDecorView().setBackgroundColor(getColor(R.color.green));

            }
        });

        blueradioBut.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getWindow().getDecorView().setBackgroundColor(getColor(R.color.blue));

            }
        });

        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                msgText.setText("");
                Log.d("my app"," Add Button Clicked");
                String name =  nameText.getText().toString();
                String yob = yearText.getText().toString();
                if (!name.isEmpty() && ! yob.isEmpty()){
                    int year = Integer.parseInt(yob);
                    if (year > 1920 && year < 2025){
                        if (stdSwitch.isChecked()){
                            msgText.setText("Welcome New Student: "+ name + ". You are "+ ( 2024 - year) + " years old");
                        }
                        else {
                            msgText.setText("Welcome " + name + ". You are " + (2024 - year) + " years old");
                        }
                        nameText.setText("");
                        yearText.setText("");

                        }else {
                        msgText.setText(R.string.error2);
                    }
                }else {
                    msgText.setText(R.string.error1);
                }
            }
        });




    }
}

File: Dec20/app/src/main/java/com/example/dec20/MainActivity.kt
package com.example.dec20

import android.annotation.SuppressLint
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

class MainActivity : AppCompatActivity() {
// variables in kotlin
    // classes
    // loops
    // functions
    // closur = lambda = inline function = function with no name
    fun myHigherOrderFunction(a: Int, b:Int, op: (Int,Int) -> Int) : Int{
        return op(a,b)
    }

    @SuppressLint("MissingInflatedId")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)

        val addbut = findViewById<Button>(R.id.add)
        val subbut = findViewById<Button>(R.id.sub)
        val timesbut = findViewById<Button>(R.id.times)
        val dividbut = findViewById<Button>(R.id.divid)
        val num1 = findViewById<EditText>(R.id.num1)
        val num2 = findViewById<EditText>(R.id.num2)
        val resultText = findViewById<TextView>(R.id.result)

        addbut.setOnClickListener{
            if ((!num1.text.toString().isEmpty()) && (!num2.text.toString().isEmpty() )){
                val n1 = num1.text.toString().toInt()
                val n2 = num2.text.toString().toInt()
                resultText.text =myHigherOrderFunction(n1, n2) {n1, n2 -> n1 + n2}.toString()
            }
        }

        subbut.setOnClickListener{
            if ((!num1.text.toString().isEmpty()) && (!num2.text.toString().isEmpty() )){
                val n1 = num1.text.toString().toInt()
                val n2 = num2.text.toString().toInt()
                resultText.text = myHigherOrderFunction(n1, n2){n1, n2 -> n1 - n2}.toString()
            }
        }

        timesbut.setOnClickListener{
            if ((!num1.text.toString().isEmpty()) && (!num2.text.toString().isEmpty() )){
                val n1 = num1.text.toString().toInt()
                val n2 = num2.text.toString().toInt()
                resultText.text = myHigherOrderFunction(n1, n2){n1, n2 -> n1 * n2}.toString()
            }
        }
        dividbut.setOnClickListener{
            if ((!num1.text.toString().isEmpty()) && (!num2.text.toString().isEmpty() )){
                val n1 = num1.text.toString().toInt()
                val n2 = num2.text.toString().toInt()
                resultText.text = myHigherOrderFunction(n1, n2){n1, n2 -> n1 / n2}.toString()
            }
        }

        val functionButton = findViewById<Button>(R.id.functionsinKotlin)

        functionButton.setOnClickListener{
                val result = sum(40,50)
                Toast.makeText(this, "The sume result of 40 and 50 is $result",Toast.LENGTH_LONG).show()
        }

        val dataclassButton = findViewById<Button>(R.id.dataclass)
        dataclassButton.setOnClickListener{
            val myFirstStudent = Student("mary",123)
            val mySecondStudent = Student("john",456)

            val allStudent = listOf<Student>(myFirstStudent,mySecondStudent)
            val allMesseges = allStudent.map{ "Welcome ${it.name} to our class, Your id is ${it.number}" }

            Toast.makeText(this, "${allMesseges[0]}",Toast.LENGTH_LONG).show()
        }

        val whenButton = findViewById<Button>(R.id.whenButton)
        whenButton.setOnClickListener {
            Toast.makeText(this, describe("Hello"),Toast.LENGTH_LONG).show()
            Toast.makeText(this, describe(1),Toast.LENGTH_LONG).show()
            Toast.makeText(this, describe(5556666111),Toast.LENGTH_LONG).show()
            Toast.makeText(this, describe(true),Toast.LENGTH_LONG).show()
            Toast.makeText(this, describe('t'),Toast.LENGTH_LONG).show()

        }

        val loopButton = findViewById<Button>(R.id.loopsinKotlin)
    loopButton.setOnClickListener {
        val numbers : List<Int> = listOf(2,5,-1,44,21)
        for (num in numbers){
            Toast.makeText(this, "number = $num",Toast.LENGTH_SHORT).show()
        }
    }

    val lambdaSum = { a: Int, b: Int -> a + b}//
    val lambdaAve = {a: Double, b: Double, c: Double -> (a + b + c)/3 }


    val lambdaButton = findViewById<Button>(R.id.lambdainKotlin)
    lambdaButton.setOnClickListener{
        val result = lambdaSum(40,50)
        Toast.makeText(this, "The Lambda result of 40 and 50 is $result",Toast.LENGTH_LONG).show()
    }

    // map , filter, count, reduce ==> Higher order functions because they have lambda as paremeter.
    // only for collections
    // map as data structure == > key - value pairs

    val filterButton = findViewById<Button>(R.id.filterFunctioninKotlin)
    filterButton.setOnClickListener{
       val numbers = arrayOf(1,4,-3,-9,2,-1,7)

       val positivNumbers =  numbers.filter{ it > 0}
        val evenNumbers = numbers.filter { it % 2 == 0 }
        // it similar to this in java

        for (num in positivNumbers){
            Toast.makeText(this, "number = $num",Toast.LENGTH_SHORT).show()
        }
    }

    val mapButton = findViewById<Button>(R.id.mapFunctioninKotlin)
    mapButton.setOnClickListener{
        val numbers = arrayOf(1,4,-3,-9,2,-1,7)
        val newNumbers = numbers.map{ it * 2}
        val stringarray = numbers.map{ "The number is $it"}
        val halfNumbers =  numbers.map{ it / 2}
        for (num in halfNumbers){
            Toast.makeText(this, "$num",Toast.LENGTH_SHORT).show()
        }
    }
    }



       fun sum(a: Int, b: Int) : Int = a + b

    fun describe(obj : Any) : String =
        when (obj){
            1           -> "One"
            "Hello"     -> "Hello"
            is Long     -> "This is Long Number"
            !is String  -> "Not A String"
            else        -> "Unknown"
        }




    }